### PR TITLE
Stop using Travis's deploy support and handle it ourselves

### PR DESCRIPTION
**Pull request by DRMacIver**
We are *so close* to this continuous deployment thing, but [the latest build](https://travis-ci.org/HypothesisWorks/hypothesis-python/jobs/224229145#L376) has Travis cheerfully going "Already up-to-date!". Thanks Travis, no it's not.

Without any further information to debug, I have adopted my traditional solution to the problem: Why do it the easy way when you can just brute force solve it yourself? We're already 99% of the way there anyway - Travis is only handling the *easy* bit.

So, this PR rips out the use of Travis's deployment logic. We now have an encrypted pypirc (for me - I should look in to creating a Hypothesis specific one) in the repo as well as a deploy key, and we then just build an sdist and invoke Twine.

Additional things this PR contains that could arguably be separated out but I don't really think should be:

* Add deploy_key to gitignore (whoops)
* Tag *after* upload rather than before. This is a much better order which is made easier to do by handling the logic ourselves, but I realise now we could and should have done it that way before.

### CODE REVIEW

**Comment by DRMacIver** [comment]
Unlike the other things I've merged by fiat I am going to wait for a review on this one because it's actually non-trivial.

File: scripts/deploy.py [hunk 1]
```diff
@@ -79,6 +81,20 @@
...
+
+    print("Now uploading to pypi.")
+
+    subprocess.check_output([
+        sys.executable, "setup.py", "sdist"
+    ])
+
+    subprocess.check_output([
+        sys.executable, "-m", "twine", "--config-file=./.pypirc",
+        "upload", "dist/*"
```
Comment (alexwlchan) [inline scripts/deploy.py]: Worth trashing the contents of the `dist` directory before you run the `sdist` step?

I can’t remember if it was Travis specifically, but I’ve had weirdness in the past when supposedly ephemeral containers got cached between runs.

Comment (DRMacIver) [inline scripts/deploy.py]: Travis specifically does this for us, but I agree that we should do it to just to be safe.

Comment (alexwlchan) [inline scripts/deploy.py]: > Travis specifically does this for us, but I agree that we should do it to just to be safe.

Yeah, I’m going to point to This Entire Pull Request as evidence that we don’t necessarily rely on Travis behaving in a sensible way. 😉

File: scripts/deploy.py [hunk 2]
```diff
@@ -79,6 +85,23 @@
...
-    print("Looks good to release! Pushing the tag now.")
+    print("Looks good to release!")
+
+    if os.path.exists(DIST):
+        shutil.rmtree(DIST)
+
+    print("Now uploading to pypi.")
+
+    subprocess.check_output([
+        sys.executable, "setup.py", "sdist"
```
Comment (alexwlchan) [inline scripts/deploy.py]: Now we have an absolute path to `DIST`, let's make sure the `sdist` step puts it there:

```python
subprocess.check_output([
    sys.executable, "setup.py", "sdist", "--dist-dir", DIST
])
```

**Comment by alexwlchan** [comment]
(I am happy for you to merge by fiat for the sake of getting the deployment code testing again)

**Comment by DRMacIver** [comment]
Hmm. The decryption is failing for some reason, so I'm not going to do that.

**Comment by DRMacIver** [comment]
I've tried just rerunning it to see if I actually stuffed something up. That produced a diff but I suspect because there's a timestamp in the encryption somewhere. Lets see what Travis does.

**Comment by DRMacIver** [comment]
It looks like there might be, uh, interesting things going on with multiple files: https://docs.travis-ci.com/user/encrypting-files/#Encrypting-multiple-files

**Comment by DRMacIver** [comment]
Right, no fucking clue what's going on there, but changing it to put both secrets in a tarfile and decrypt the tarfile seems to have worked. https://travis-ci.org/HypothesisWorks/hypothesis-python/jobs/224245730#L168 shows we're getting both files out of it. I'm now going to admin fiat this.

**Pull request merged**